kafka-rest/src/main/java/io/confluent/kafkarest/controllers/ConsumerOffsetsDaoImpl.java

```diff
@@ -0,0 +1,354 @@
+        .timeoutMs(consumerMetadataTimeout);
+
+    Map<TopicPartition, OffsetSpec> latestOffsetSpecs = fetchedCurrentOffsets.keySet().stream()
+        .collect(Collectors.toMap(Function.identity(), tp -> OffsetSpec.latest()));
+
+    ListOffsetsResult latestOffsetResult = kafkaAdminClient.listOffsets(
+        latestOffsetSpecs, listOffsetsOptions);
+    Map<TopicPartition, ListOffsetsResultInfo> latestOffsets = latestOffsetResult.all()
+        .get(consumerMetadataTimeout, TimeUnit.MILLISECONDS);
+
+    Map<TopicPartition, MemberId> tpMemberIds = new HashMap<>();
+
+    for (MemberDescription memberDesc : cgDesc.members()) {
+      for (TopicPartition tp : memberDesc.assignment().topicPartitions()) {
+        MemberId memberId =
+            MemberId.builder()
+                .setConsumerId(memberDesc.consumerId())
+                .setClientId(memberDesc.clientId())
+                .setInstanceId(memberDesc.groupInstanceId())
+                .build();
+        tpMemberIds.put(tp, memberId);
+      }
+    }
+
+    for (TopicPartition tp : fetchedCurrentOffsets.keySet()) {
+      // look up consumer id from map if not part of a simple consumer group
+      MemberId memberId = tpMemberIds.getOrDefault(
+          tp, MemberId.builder()
+              .setConsumerId("")
+              .setClientId("")
+              .setInstanceId(Optional.empty())
+              .build());
+
+      long currentOffset = getCurrentOffset(fetchedCurrentOffsets, tp);
+      long latestOffset = getOffset(latestOffsets, tp);
+      if (currentOffset < 0 || latestOffset < 0) {
+        // log.debug("invalid offsets for topic={} consumerId={} current={} latest={}",
+        //     tp.topic(),
+        //     consumerId,
+        //     currentOffset,
+        //     latestOffset);
+        continue;
+      }
+      ConsumerLag consumerLag =
+          ConsumerLag.builder()
+              .setClusterId(clusterId)
+              .setConsumerGroupId(consumerGroupId)
+              .setTopicName(tp.topic())
+              .setPartitionId(tp.partition())
+              .setConsumerId(memberId.getConsumerId())
+              .setInstanceId(memberId.getInstanceId().orElse(null))
+              .setClientId(memberId.getClientId())
+              .setCurrentOffset(currentOffset)
+              .setLogEndOffset(latestOffset)
+              .build();
+      consumerLags.add(consumerLag);
+    }
+    return consumerLags;
+  }
+
+  ConsumerGroupLag.Builder getConsumerGroupOffsets(
+      ConsumerGroupDescription cgDesc,
+      Map<TopicPartition, OffsetAndMetadata> fetchedCurrentOffsets,
+      Map<TopicPartition, ListOffsetsResultInfo> latestOffsets
+  ) {
+    ConsumerGroupLag.Builder cgOffsets =
+        ConsumerGroupLag.builder().setConsumerGroupId(cgDesc.groupId());
+
+    // build map of topic partition -> consumer, client, instance ids
+    // Map<TopicPartition, String> tpConsumerIds = new HashMap<>();
+    // Map<TopicPartition, String> tpClientIds = new HashMap<>();
```

Comment:
Comments kept temporarily for reference, blueway used two separate Maps to keep track of topic-partition to consumerId and clientId mappings. I created MemberId to keep track of these mappings (with the addition of instanceIds) in a single Map